### PR TITLE
Adding back-off cap of 5 seconds for 429 and skipping metadata calls for queries with partitioning strategy Restrictive

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

This PR contains two small optimizations

1. SDK + Spark - currently it is possible that the Cosmos DB backend will ask for a back-off time of significantly larger than 5 seconds (we have seen in logs up to 260 seconds) for 429/3200 (throttling because provisioned throughput is exceeded) if a replica has previously returned multiple consecutive 429s already. This PR changes the logic to enforce a n upper boundary of the back-off time of 5 seconds. This change has been vetted with Prashant (owning resource governance on the service)
2. Spark - for queries with partitioning strategy restrictive we currently still retrieve the metadata (min LSN, max LSN, document count, total document size) - this information is used to calculate the number of Spark partitions to create for non-restrictive partitioning strategies. For queries with restrictive partitioning the metadata is technically not necessary - so this PR adds an optimization to skip the metadata I/O calls when not strictly necessary.

# All SDK Contribution checklist:
- [X] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala [hunk 1]
```diff
@@ -661,14 +680,32 @@ private object CosmosPartitionPlanner extends BasicLoggingTrait {
...
             normalizedRange =>
-              PartitionMetadataCache.apply(
-                userConfig,
-                cosmosClientConfig,
-                cosmosClientStateHandle,
-                cosmosContainerConfig,
-                normalizedRange,
-                maxStaleness
-            ))
+              if (isChangeFeed != partitionConfig.partitioningStrategy != PartitioningStrategies.Restrictive) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala:683]: !isChangeFeed ?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala]: Fixed the typo - good catch. Also added a comment explaining the logic

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks for the quick fix and solution 👍

**Review by kushagraThapar — COMMENTED** [review]
@FabianMeiswinkel  - I see we have not updated the other places where we use retry-after duration like, `BatchExecUtils`, and in `BatchResponseParser` #getRetryAfterDuration() 
Any reason for not changing these?

**Comment by FabianMeiswinkel** [comment]
> @FabianMeiswinkel - I see we have not updated the other places where we use retry-after duration like, `BatchExecUtils`, and in `BatchResponseParser` #getRetryAfterDuration() Any reason for not changing these?

Good catch - I initially thought it isn't necessary to also cap the back-off time here, because when a BatchResponse overall result is 429 it gets ultimately converted into CosmosException (and we would have the cap there). But for Bulk it is possible that 429 is just one of the inner results - and there applying the cap would be needed. So I changed this to also always cap the BacthResult back-off.

**Review by kushagraThapar — APPROVED** [review]

**Pull request merged**